File: java/src/main/java/com/authdog/AuthdogClient.java
package com.authdog;

import com.authdog.exceptions.AuthenticationException;
import com.authdog.exceptions.ApiException;
import com.authdog.resources.EnvironmentsResource;
import com.authdog.resources.GroupsResource;
import com.authdog.resources.OrganizationsResource;
import com.authdog.resources.ProjectsResource;
import com.authdog.resources.TenantsResource;
import com.authdog.resources.UsersResource;
import com.authdog.types.Probe;
import com.authdog.types.UserInfoResponse;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.HttpUrl;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Main client for interacting with Authdog API.
 */
public class AuthdogClient implements AutoCloseable {
    /**
     * Default timeout in milliseconds.
     */
    private static final int DEFAULT_TIMEOUT_MS = 10000;

    /**
     * HTTP 200 status code.
     */
    private static final int HTTP_OK = 200;

    /**
     * HTTP 400 status code.
     */
    private static final int HTTP_BAD_REQUEST = 400;

    /**
     * HTTP 401 status code.
     */
    private static final int HTTP_UNAUTHORIZED = 401;

    /**
     * HTTP 500 status code.
     */
    private static final int HTTP_INTERNAL_SERVER_ERROR = 500;

    /**
     * JSON media type.
     */
    private static final MediaType JSON_MEDIA =
            MediaType.get("application/json");

    /**
     * Empty JSON request body.
     */
    private static final RequestBody EMPTY_BODY =
            RequestBody.create("", JSON_MEDIA);

    /**
     * HTTP client.
     */
    private final OkHttpClient httpClient;

    /**
     * Base URL.
     */
    private final String baseUrl;

    /**
     * API key.
     */
    private final String apiKey;

    /**
     * Object mapper.
     */
    private final ObjectMapper objectMapper;

    /**
     * Organizations resource.
     */
    private final OrganizationsResource organizations;

    /**
     * Tenants resource.
     */
    private final TenantsResource tenants;

    /**
     * Projects resource.
     */
    private final ProjectsResource projects;

    /**
     * Environments resource.
     */
    private final EnvironmentsResource environments;

    /**
     * Users resource.
     */
    private final UsersResource users;

    /**
     * Groups resource.
     */
    private final GroupsResource groups;

    /**
     * Initialize the Authdog client.
     * @param baseUrlParam The base URL of the Authdog API
     */
    public AuthdogClient(final String baseUrlParam) {
        this(baseUrlParam, null, DEFAULT_TIMEOUT_MS);
    }

    /**
     * Initialize the Authdog client with API key.
     * @param baseUrlParam The base URL of the Authdog API
     * @param apiKeyParam Optional API key for authentication
     */
    public AuthdogClient(final String baseUrlParam, final String apiKeyParam) {
        this(baseUrlParam, apiKeyParam, DEFAULT_TIMEOUT_MS);
    }

    /**
     * Initialize the Authdog client with custom timeout.
     * @param baseUrlParam The base URL of the Authdog API
     * @param apiKeyParam Optional API key for authentication
     * @param timeoutMsParam Timeout in milliseconds
     */
    public AuthdogClient(final String baseUrlParam, final String apiKeyParam,
                        final int timeoutMsParam) {
        this.baseUrl = baseUrlParam.endsWith("/")
                ? baseUrlParam.substring(0, baseUrlParam.length() - 1)
                : baseUrlParam;
        this.apiKey = apiKeyParam;
        this.objectMapper = new ObjectMapper();

        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(timeoutMsParam, TimeUnit.MILLISECONDS)
                .readTimeout(timeoutMsParam, TimeUnit.MILLISECONDS)
                .writeTimeout(timeoutMsParam, TimeUnit.MILLISECONDS)
                .build();
        this.organizations = new OrganizationsResource(this);
        this.tenants = new TenantsResource(this);
        this.projects = new ProjectsResource(this);
        this.environments = new EnvironmentsResource(this);
        this.users = new UsersResource(this);
        this.groups = new GroupsResource(this);
    }

    /**
     * Management API key used as a Bearer credential.
     * Userinfo uses the access token argument instead.
     * @return API key or null
     */
    public String getApiKey() {
        return apiKey;
    }

    /**
     * Organization management helpers.
     * @return organizations resource
     */
    public OrganizationsResource organizations() {
        return organizations;
    }

    /**
     * Tenant management helpers.
     * @return tenants resource
     */
    public TenantsResource tenants() {
        return tenants;
    }

    /**
     * Project management helpers.
     * @return projects resource
     */
    public ProjectsResource projects() {
        return projects;
    }

    /**
     * Environment management helpers.
     * @return environments resource
     */
    public EnvironmentsResource environments() {
        return environments;
    }

    /**
     * Directory user helpers.
     * @return users resource
     */
    public UsersResource users() {
        return users;
    }

    /**
     * Directory group helpers.
     * @return groups resource
     */
    public GroupsResource groups() {
        return groups;
    }

    /**
     * Liveness probe. Public; works without a management credential.
     * @return probe result
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public Probe health() throws AuthenticationException, ApiException {
        return request("GET", "/v1/health", null, null, Probe.class);
    }

    /**
     * Send a JSON management request.
     * @param <T> response type
     * @param methodParam HTTP method
     * @param pathParam request path beginning with /
     * @param bodyParam optional JSON body
     * @param queryParam optional query parameters
     * @param typeParam Jackson type
     * @return parsed response
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public <T> T request(final String methodParam, final String pathParam,
                         final Object bodyParam,
                         final Map<String, String> queryParam,
                         final Class<T> typeParam)
            throws AuthenticationException, ApiException {
        final Request.Builder builder = new Request.Builder()
                .url(buildUrl(pathParam, queryParam))
                .addHeader("Content-Type", "application/json")
                .addHeader("User-Agent", "authdog-java-sdk/0.1.0");
        if (apiKey != null) {
            builder.header("Authorization", "Bearer " + apiKey);
        }
        applyMethod(builder, methodParam, encodeBody(bodyParam));

        try (Response response = httpClient.newCall(builder.build())
                .execute()) {
            final ResponseBody body = response.body();
            String responseBody = body != null ? body.string() : "";
            throwIfUnsuccessful(response.code(), responseBody);
            if (responseBody.isEmpty()) {
                responseBody = "{}";
            }
            try {
                return objectMapper.readValue(responseBody, typeParam);
            } catch (Exception e) {
                throw new ApiException("Failed to parse response: "
                        + e.getMessage(), e);
            }
        } catch (IOException e) {
            throw new ApiException("Request failed: " + e.getMessage(), e);
        }
    }

    /**
     * Get user information using an access token.
     * @param accessTokenParam The access token for authentication
     * @return UserInfoResponse containing user information
     * @throws AuthenticationException When authentication fails
     * @throws ApiException When API request fails
     */
    public UserInfoResponse getUserInfo(final String accessTokenParam)
            throws AuthenticationException, ApiException {
        String url = baseUrl + "/v1/userinfo";

        Request.Builder requestBuilder = new Request.Builder()
                .url(url)
                .addHeader("Content-Type", "application/json")
                .addHeader("User-Agent", "authdog-java-sdk/0.1.0")
                .header("Authorization", "Bearer " + accessTokenParam);

        Request request = requestBuilder.build();

        try (Response response = httpClient.newCall(request).execute()) {
            ResponseBody body = response.body();
            String responseBody = body != null ? body.string() : "";

            if (response.code() == HTTP_UNAUTHORIZED) {
                throw new AuthenticationException(
                        "Unauthorized - invalid or expired token");
            }

            if (response.code() == HTTP_INTERNAL_SERVER_ERROR) {
                try {
                    // Try to parse error response
                    var errorNode = objectMapper.readTree(responseBody);
                    if (errorNode.has("error")) {
                        String errorMessage = errorNode.get("error").asText();
                        if ("GraphQL query failed".equals(errorMessage)) {
                            throw new ApiException("GraphQL query failed");
                        } else if ("Failed to fetch user info"
                                .equals(errorMessage)) {
                            throw new ApiException("Failed to fetch user info");
                        }
                    }
                } catch (IOException e) {
                    // Ignore JSON parsing errors for error responses
                }
                throw new ApiException("HTTP error 500: " + responseBody);
            }

            if (response.code() != HTTP_OK) {
                throw new ApiException("HTTP error " + response.code()
                        + ": " + responseBody);
            }

            try {
                return objectMapper.readValue(responseBody,
                        UserInfoResponse.class);
            } catch (Exception e) {
                throw new ApiException("Failed to parse response: "
                        + e.getMessage(), e);
            }

        } catch (IOException e) {
            throw new ApiException("Request failed: " + e.getMessage(), e);
        }
    }
    /**
     * Close the HTTP client.
     */
    @Override
    public void close() {
        httpClient.dispatcher().executorService().shutdown();
        httpClient.connectionPool().evictAll();
    }

    /**
     * Build the request URL.
     * @param pathParam request path
     * @param queryParam optional query parameters
     * @return absolute URL
     */
    private HttpUrl buildUrl(final String pathParam,
                             final Map<String, String> queryParam) {
        final HttpUrl parsed = HttpUrl.parse(baseUrl + pathParam);
        if (parsed == null) {
            throw new ApiException("Request failed: invalid URL");
        }
        if (queryParam == null || queryParam.isEmpty()) {
            return parsed;
        }
        final HttpUrl.Builder urlBuilder = parsed.newBuilder();
        for (final Map.Entry<String, String> entry
                : queryParam.entrySet()) {
            if (entry.getValue() != null) {
                urlBuilder.addQueryParameter(entry.getKey(),
                        entry.getValue());
            }
        }
        return urlBuilder.build();
    }

    /**
     * Encode an optional JSON body.
     * @param bodyParam request body
     * @return OkHttp body
     */
    private RequestBody encodeBody(final Object bodyParam) {
        if (bodyParam == null) {
            return EMPTY_BODY;
        }
        try {
            return RequestBody.create(
                    objectMapper.writeValueAsString(bodyParam),
                    JSON_MEDIA);
        } catch (IOException e) {
            throw new ApiException("Request failed: " + e.getMessage(), e);
        }
    }

    /**
     * Apply the HTTP method to the request builder.
     * @param builderParam request builder
     * @param methodParam HTTP method
     * @param bodyParam encoded body
     */
    private void applyMethod(final Request.Builder builderParam,
                             final String methodParam,
                             final RequestBody bodyParam) {
        switch (methodParam) {
            case "GET":
                builderParam.get();
                break;
            case "POST":
                builderParam.post(bodyParam);
                break;
            case "PUT":
                builderParam.put(bodyParam);
                break;
            case "PATCH":
                builderParam.patch(bodyParam);
                break;
            case "DELETE":
                builderParam.delete();
                break;
            default:
                throw new ApiException("Request failed: unsupported HTTP "
                        + "method");
        }
    }

    /**
     * Map unsuccessful HTTP responses onto SDK exceptions.
     * @param codeParam HTTP status
     * @param responseBodyParam response body
     */
    private void throwIfUnsuccessful(final int codeParam,
                                     final String responseBodyParam) {
        if (codeParam == HTTP_UNAUTHORIZED) {
            throw new AuthenticationException(
                    "Unauthorized - invalid or expired token");
        }
        if (codeParam >= HTTP_BAD_REQUEST) {
            throw new ApiException("HTTP error " + codeParam + ": "
                    + extractErrorText(responseBodyParam));
        }
    }

    /**
     * Prefer a JSON {@code error} field when present.
     * @param responseBodyParam response body
     * @return error text
     */
    private String extractErrorText(final String responseBodyParam) {
        try {
            final JsonNode node = objectMapper.readTree(responseBodyParam);
            if (node != null && node.has("error")
                    && !node.get("error").isNull()) {
                return node.get("error").asText();
            }
        } catch (IOException e) {
            // Use the raw body when the error payload is not JSON.
        }
        return responseBodyParam;
    }
}

File: csharp/AuthdogClient.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Authdog.Exceptions;
using Authdog.Types;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Main client for interacting with Authdog API
    /// </summary>
    public class AuthdogClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly string _baseUrl;
        private readonly string? _apiKey;
        private readonly string? _environmentSecret;
        private readonly string? _scimToken;
        private readonly string? _hrisToken;
        private bool _disposed = false;

        /// <summary>
        /// Timeout applied to the owned HTTP client. Injected clients keep their own timeout.
        /// </summary>
        public TimeSpan Timeout { get; }

        /// <summary>
        /// Optional management Bearer credential. Userinfo still uses the access-token argument.
        /// </summary>
        public string? ApiKey => _apiKey;

        /// <summary>
        /// Optional <c>adenv_</c> secret for AuthZEN evaluation/search and MCP runtime.
        /// </summary>
        public string? EnvironmentSecret => _environmentSecret;

        /// <summary>
        /// Optional <c>adscim_</c> token for <c>/v1/scim/v2</c>.
        /// </summary>
        public string? ScimToken => _scimToken;

        /// <summary>
        /// Optional <c>adhris_</c> token for <c>/v1/hris/v1</c>.
        /// </summary>
        public string? HrisToken => _hrisToken;

        public OrganizationsResource Organizations { get; }

        public TenantsResource Tenants { get; }

        public ProjectsResource Projects { get; }

        public EnvironmentsResource Environments { get; }

        public UsersResource Users { get; }

        public GroupsResource Groups { get; }

        public RbacResource Rbac { get; }

        public AuditResource Audit { get; }

        public EventsResource Events { get; }

        public WebhooksResource Webhooks { get; }

        public NotificationChannelsResource NotificationChannels { get; }

        public ServiceAccountsResource ServiceAccounts { get; }

        public PersonalAccessTokensResource PersonalAccessTokens { get; }

        public ApiSecretsResource ApiSecrets { get; }

        public AuthzenResource Authzen { get; }

        public ScimResource Scim { get; }

        public HrisResource Hris { get; }

        public McpResource Mcp { get; }

        public OtelResource Otel { get; }

        public OidcClientsResource OidcClients { get; }

        public ActionsResource Actions { get; }

        public AddonsResource Addons { get; }

        public BillingResource Billing { get; }

        public SettingsResource Settings { get; }

        public ElevateResource Elevate { get; }

        public EmailProvidersResource EmailProviders { get; }

        public FeatureFlagsResource FeatureFlags { get; }

        public FormsResource Forms { get; }

        public ProvisioningTokensResource ProvisioningTokens { get; }

        public ImpersonationResource Impersonation { get; }

        public PortalResource Portal { get; }

        public SecurityResource Security { get; }

        public ThreatsResource Threats { get; }

        public VanityDomainsResource VanityDomains { get; }

        public WidgetsResource Widgets { get; }

        /// <summary>
        /// Initialize the Authdog client
        /// </summary>
        /// <param name="baseUrl">The base URL of the Authdog API</param>
        /// <param name="apiKey">Optional management Bearer credential; unused on userinfo</param>
        /// <param name="httpClient">Optional custom HttpClient instance</param>
        /// <param name="timeout">Timeout for an owned HttpClient (default 10 seconds)</param>
        /// <param name="environmentSecret">Optional <c>adenv_</c> secret for AuthZEN and MCP runtime</param>
        /// <param name="scimToken">Optional <c>adscim_</c> token for SCIM</param>
        /// <param name="hrisToken">Optional <c>adhris_</c> token for HRIS</param>
        public AuthdogClient(
            string baseUrl,
            string? apiKey = null,
            HttpClient? httpClient = null,
            TimeSpan? timeout = null,
            string? environmentSecret = null,
            string? scimToken = null,
            string? hrisToken = null)
        {
            _baseUrl = baseUrl.TrimEnd('/');
            _apiKey = apiKey;
            _environmentSecret = environmentSecret;
            _scimToken = scimToken;
            _hrisToken = hrisToken;
            Timeout = timeout ?? TimeSpan.FromSeconds(10);
            _httpClient = httpClient ?? new HttpClient { Timeout = Timeout };

            if (!_httpClient.DefaultRequestHeaders.Contains("User-Agent"))
            {
                _httpClient.DefaultRequestHeaders.Add("User-Agent", "authdog-csharp-sdk/0.1.0");
            }

            Organizations = new OrganizationsResource(this);
            Tenants = new TenantsResource(this);
            Projects = new ProjectsResource(this);
            Environments = new EnvironmentsResource(this);
            Users = new UsersResource(this);
            Groups = new GroupsResource(this);
            Rbac = new RbacResource(this);
            Audit = new AuditResource(this);
            Events = new EventsResource(this);
            Webhooks = new WebhooksResource(this);
            NotificationChannels = new NotificationChannelsResource(this);
            ServiceAccounts = new ServiceAccountsResource(this);
            PersonalAccessTokens = new PersonalAccessTokensResource(this);
            ApiSecrets = new ApiSecretsResource(this);
            Authzen = new AuthzenResource(this);
            Scim = new ScimResource(this);
            Hris = new HrisResource(this);
            Mcp = new McpResource(this);
            Otel = new OtelResource(this);
            OidcClients = new OidcClientsResource(this);
            Actions = new ActionsResource(this);
            Addons = new AddonsResource(this);
            Billing = new BillingResource(this);
            Settings = new SettingsResource(this);
            Elevate = new ElevateResource(this);
            EmailProviders = new EmailProvidersResource(this);
            FeatureFlags = new FeatureFlagsResource(this);
            Forms = new FormsResource(this);
            ProvisioningTokens = new ProvisioningTokensResource(this);
            Impersonation = new ImpersonationResource(this);
            Portal = new PortalResource(this);
            Security = new SecurityResource(this);
            Threats = new ThreatsResource(this);
            VanityDomains = new VanityDomainsResource(this);
            Widgets = new WidgetsResource(this);
        }

        /// <summary>
        /// Environment-scoped management path prefix.
        /// </summary>
        internal static string Env(string tenantId, string environmentId) =>
            $"/v1/tenants/{tenantId}/environments/{environmentId}";

        /// <summary>
        /// Build a query map, dropping null values.
        /// </summary>
        internal static IDictionary<string, string?>? Params(params (string Key, object? Value)[] pairs)
        {
            var query = new Dictionary<string, string?>();
            foreach (var (key, value) in pairs)
            {
                if (value is null)
                {
                    continue;
                }

                query[key] = Convert.ToString(value, CultureInfo.InvariantCulture);
            }

            return query.Count == 0 ? null : query;
        }

        /// <summary>
        /// Convert a Dictionary or object into query parameters, dropping null values.
        /// </summary>
        internal static IDictionary<string, string?>? QueryFrom(object? query)
        {
            if (query == null)
            {
                return null;
            }

            if (query is IDictionary<string, string?> typedNullable)
            {
                var copy = new Dictionary<string, string?>();
                foreach (var pair in typedNullable)
                {
                    if (pair.Value != null)
                    {
                        copy[pair.Key] = pair.Value;
                    }
                }

                return copy.Count == 0 ? null : copy;
            }

            if (query is IDictionary dictionary)
            {
                var copy = new Dictionary<string, string?>();
                foreach (DictionaryEntry entry in dictionary)
                {
                    if (entry.Key is null || entry.Value is null)
                    {
                        continue;
                    }

                    var key = Convert.ToString(entry.Key, CultureInfo.InvariantCulture);
                    if (string.IsNullOrEmpty(key))
                    {
                        continue;
                    }

                    copy[key] = Convert.ToString(entry.Value, CultureInfo.InvariantCulture);
                }

                return copy.Count == 0 ? null : copy;
            }

            var obj = query as JObject ?? JObject.FromObject(query);
            var result = new Dictionary<string, string?>();
            foreach (var property in obj.Properties())
            {
                if (property.Value.Type is JTokenType.Null or JTokenType.Undefined)
                {
                    continue;
                }

                result[property.Name] = property.Value.Type == JTokenType.String
                    ? property.Value.Value<string>()
                    : property.Value.ToString(Formatting.None);
            }

            return result.Count == 0 ? null : result;
        }

        /// <summary>
        /// Send a JSON management request. Constructor apiKey is sent as Bearer unless
        /// <paramref name="accessToken"/> is provided. <paramref name="omitAuth"/> skips
        /// Authorization even when ApiKey is set. Does not set DefaultRequestHeaders.
        /// </summary>
        public async Task<T> RequestAsync<T>(
            HttpMethod method,
            string path,
            object? body = null,
            IDictionary<string, string?>? query = null,
            string? accessToken = null,
            bool omitAuth = false)
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(AuthdogClient));

            var request = new HttpRequestMessage(method, BuildRequestUri(path, query));
            if (!omitAuth)
            {
                var token = accessToken ?? _apiKey;
                if (token != null)
                {
                    request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
                }
            }

            if (body != null)
            {
                var json = JsonConvert.SerializeObject(body);
                request.Content = new StringContent(json, Encoding.UTF8, "application/json");
            }

            try
            {
                var response = await _httpClient.SendAsync(request);
                var content = await response.Content.ReadAsStringAsync();

                if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
                {
                    throw new AuthenticationException("Unauthorized - invalid or expired token");
                }

                if (!response.IsSuccessStatusCode)
                {
                    var errorText = content;
                    try
                    {
                        var payload = JsonConvert.DeserializeObject<JObject>(content);
                        var error = payload?["error"];
                        if (error != null && error.Type != JTokenType.Null)
                        {
                            errorText = error.Type == JTokenType.String ? error.ToString() : error.ToString();
                        }
                    }
                    catch (JsonException)
                    {
                    }

                    throw new ApiException($"HTTP error {(int)response.StatusCode}: {errorText}", (int)response.StatusCode);
                }

                if (string.IsNullOrWhiteSpace(content))
                {
                    if (typeof(T) == typeof(JObject))
                    {
                        return (T)(object)new JObject();
                    }

                    return JsonConvert.DeserializeObject<T>("{}")
                        ?? throw new ApiException("Failed to parse response: invalid JSON");
                }

                try
                {
                    return JsonConvert.DeserializeObject<T>(content)
                        ?? throw new ApiException("Failed to parse response: invalid JSON");
                }
                catch (JsonException ex)
                {
                    throw new ApiException("Failed to parse response: invalid JSON", ex);
                }
            }
            catch (AuthenticationException)
            {
                throw;
            }
            catch (ApiException)
            {
                throw;
            }
            catch (HttpRequestException ex)
            {
                throw new ApiException($"Request failed: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Synchronous version of RequestAsync
        /// </summary>
        public T Request<T>(
            HttpMethod method,
            string path,
            object? body = null,
            IDictionary<string, string?>? query = null,
            string? accessToken = null,
            bool omitAuth = false)
        {
            return RequestAsync<T>(method, path, body, query, accessToken, omitAuth).GetAwaiter().GetResult();
        }

        /// <summary>
        /// Liveness probe. Public; works without a management credential.
        /// </summary>
        public Task<Probe> HealthAsync() => RequestAsync<Probe>(HttpMethod.Get, "/v1/health");

        /// <summary>
        /// Synchronous version of HealthAsync
        /// </summary>
        public Probe Health() => HealthAsync().GetAwaiter().GetResult();

        private string BuildRequestUri(string path, IDictionary<string, string?>? query)
        {
            if (!path.StartsWith("/"))
            {
                path = "/" + path;
            }

            var url = $"{_baseUrl}{path}";
            if (query == null || query.Count == 0)
            {
                return url;
            }

            var parts = new List<string>();
            foreach (var pair in query)
            {
                if (pair.Value == null)
                {
                    continue;
                }

                parts.Add($"{Uri.EscapeDataString(pair.Key)}={Uri.EscapeDataString(pair.Value)}");
            }

            return parts.Count == 0 ? url : $"{url}?{string.Join("&", parts)}";
        }

        /// <summary>
        /// Get user information using an access token
        /// </summary>
        /// <param name="accessToken">The access token for authentication</param>
        /// <returns>UserInfoResponse containing user information</returns>
        /// <exception cref="AuthenticationException">When authentication fails</exception>
        /// <exception cref="ApiException">When API request fails</exception>
        public async Task<UserInfoResponse> GetUserInfoAsync(string accessToken)
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(AuthdogClient));

            var request = new HttpRequestMessage(HttpMethod.Get, $"{_baseUrl}/v1/userinfo");
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", accessToken);

            try
            {
                var response = await _httpClient.SendAsync(request);
                var content = await response.Content.ReadAsStringAsync();

                if (response.IsSuccessStatusCode)
                {
                    return JsonConvert.DeserializeObject<UserInfoResponse>(content) ?? new UserInfoResponse();
                }

                if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
                {
                    throw new AuthenticationException("Unauthorized - invalid or expired token");
                }

                if (response.StatusCode == System.Net.HttpStatusCode.InternalServerError)
                {
                    try
                    {
                        var errorData = JsonConvert.DeserializeObject<dynamic>(content);
                        if (errorData?.error != null)
                        {
                            string errorMessage = errorData.error.ToString();
                            if (errorMessage == "GraphQL query failed")
                            {
                                throw new ApiException("GraphQL query failed");
                            }
                            else if (errorMessage == "Failed to fetch user info")
                            {
                                throw new ApiException("Failed to fetch user info");
                            }
                        }
                    }
                    catch (JsonException)
                    {
                        // Ignore JSON parsing errors for error responses
                    }
                }

                throw new ApiException($"HTTP error {(int)response.StatusCode}: {content}");
            }
            catch (HttpRequestException ex)
            {
                throw new ApiException($"Request failed: {ex.Message}", ex);
            }
            catch (TaskCanceledException ex)
            {
                throw new ApiException($"Request timeout: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Synchronous version of GetUserInfoAsync
        /// </summary>
        /// <param name="accessToken">The access token for authentication</param>
        /// <returns>UserInfoResponse containing user information</returns>
        public UserInfoResponse GetUserInfo(string accessToken)
        {
            return GetUserInfoAsync(accessToken).GetAwaiter().GetResult();
        }

        /// <summary>
        /// Dispose the HTTP client
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!_disposed && disposing)
            {
                _httpClient?.Dispose();
                _disposed = true;
            }
        }
    }
}
